In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt

In [41]:
# On import les données d'entraînement.
data = pd.read_csv("train.csv", index_col=(0))
data.head()

species   margin1   margin2   margin3   margin4   margin5  \
id                                                                            
1             Acer_Opalus  0.007812  0.023438  0.023438  0.003906  0.011719   
2   Pterocarya_Stenoptera  0.005859  0.000000  0.031250  0.015625  0.025391   
3    Quercus_Hartwissiana  0.005859  0.009766  0.019531  0.007812  0.003906   
5         Tilia_Tomentosa  0.000000  0.003906  0.023438  0.005859  0.021484   
6      Quercus_Variabilis  0.005859  0.003906  0.048828  0.009766  0.013672   

     margin6   margin7  margin8   margin9  ...  texture55  texture56  \
id                                         ...                         
1   0.009766  0.027344      0.0  0.001953  ...   0.007812   0.000000   
2   0.001953  0.019531      0.0  0.000000  ...   0.000977   0.000000   
3   0.005859  0.068359      0.0  0.000000  ...   0.154300   0.000000   
5   0.019531  0.023438      0.0  0.013672  ...   0.000000   0.000977   
6   0.015625  0.005859      0.0  0.000000  ...   0.096680   0.000000   

    texture57  texture58  texture59  texture60  texture61  texture62  \
id                                                                     
1    0.002930   0.002930   0.035156        0.0        0.0   0.004883   
2    0.000000   0.000977   0.023438        0.0        0.0   0.000977   
3    0.005859   0.000977   0.007812        0.0        0.0   0.000000   
5    0.000000   0.000000   0.020508        0.0        0.0   0.017578   
6    0.021484   0.000000   0.000000        0.0        0.0   0.000000   

    texture63  texture64  
id                        
1    0.000000   0.025391  
2    0.039062   0.022461  
3    0.020508   0.002930  
5    0.000000   0.047852  
6    0.000000   0.031250  

[5 rows x 193 columns]

In [42]:
data.shape

(990, 193)

In [43]:
# On importe les données de test.
data_validation = pd.read_csv("test.csv", index_col=(0))
data_validation.head()

margin1   margin2   margin3   margin4   margin5   margin6   margin7  \
id                                                                         
4   0.019531  0.009766  0.078125  0.011719  0.003906  0.015625  0.005859   
7   0.007812  0.005859  0.064453  0.009766  0.003906  0.013672  0.007812   
9   0.000000  0.000000  0.001953  0.021484  0.041016  0.000000  0.023438   
12  0.000000  0.000000  0.009766  0.011719  0.017578  0.000000  0.003906   
13  0.001953  0.000000  0.015625  0.009766  0.039062  0.000000  0.009766   

    margin8   margin9  margin10  ...  texture55  texture56  texture57  \
id                               ...                                    
4       0.0  0.005859  0.023438  ...   0.006836   0.000000   0.015625   
7       0.0  0.033203  0.023438  ...   0.000000   0.000000   0.006836   
9       0.0  0.011719  0.005859  ...   0.128910   0.000000   0.000977   
12      0.0  0.003906  0.001953  ...   0.012695   0.015625   0.002930   
13      0.0  0.005859  0.000000  ...   0.000000   0.042969   0.016602   

    texture58  texture59  texture60  texture61  texture62  texture63  \
id                                                                     
4    0.000977   0.015625        0.0        0.0   0.000000   0.003906   
7    0.001953   0.013672        0.0        0.0   0.000977   0.037109   
9    0.000000   0.000000        0.0        0.0   0.015625   0.000000   
12   0.036133   0.013672        0.0        0.0   0.089844   0.000000   
13   0.010742   0.041016        0.0        0.0   0.007812   0.009766   

    texture64  
id             
4    0.053711  
7    0.044922  
9    0.000000  
12   0.008789  
13   0.007812  

[5 rows x 192 columns]

In [44]:
data_validation.shape

(594, 192)

On voit qu'il manque une colonne dans le jeu de validation. Trouvons de quelle colonne il s'agit.

In [45]:
set(data.columns).symmetric_difference(set(data_validation.columns))

{'species'}

## Tour d'horizon des données

On va s'intéresser exclusivement au jeu data dans ce TP.

In [46]:
data["species"].unique().shape

(99,)

In [47]:
data.groupby("species").count()

margin1  margin2  margin3  margin4  margin5  \
species                                                                     
Acer_Capillipes                    10       10       10       10       10   
Acer_Circinatum                    10       10       10       10       10   
Acer_Mono                          10       10       10       10       10   
Acer_Opalus                        10       10       10       10       10   
Acer_Palmatum                      10       10       10       10       10   
...                               ...      ...      ...      ...      ...   
Tilia_Tomentosa                    10       10       10       10       10   
Ulmus_Bergmanniana                 10       10       10       10       10   
Viburnum_Tinus                     10       10       10       10       10   
Viburnum_x_Rhytidophylloides       10       10       10       10       10   
Zelkova_Serrata                    10       10       10       10       10   

                              margin6  margin7  margin8  margin9  margin10  \
species                                                                      
Acer_Capillipes                    10       10       10       10        10   
Acer_Circinatum                    10       10       10       10        10   
Acer_Mono                          10       10       10       10        10   
Acer_Opalus                        10       10       10       10        10   
Acer_Palmatum                      10       10       10       10        10   
...                               ...      ...      ...      ...       ...   
Tilia_Tomentosa                    10       10       10       10        10   
Ulmus_Bergmanniana                 10       10       10       10        10   
Viburnum_Tinus                     10       10       10       10        10   
Viburnum_x_Rhytidophylloides       10       10       10       10        10   
Zelkova_Serrata                    10       10       10       10        10   

                              ...  texture55  texture56  texture57  texture58  \
species                       ...                                               
Acer_Capillipes               ...         10         10         10         10   
Acer_Circinatum               ...         10         10         10         10   
Acer_Mono                     ...         10         10         10         10   
Acer_Opalus                   ...         10         10         10         10   
Acer_Palmatum                 ...         10         10         10         10   
...                           ...        ...        ...        ...        ...   
Tilia_Tomentosa               ...         10         10         10         10   
Ulmus_Bergmanniana            ...         10         10         10         10   
Viburnum_Tinus                ...         10         10         10         10   
Viburnum_x_Rhytidophylloides  ...         10         10         10         10   
Zelkova_Serrata               ...         10         10         10         10   

                              texture59  texture60  texture61  texture62  \
species                                                                    
Acer_Capillipes                      10         10         10         10   
Acer_Circinatum                      10         10         10         10   
Acer_Mono                            10         10         10         10   
Acer_Opalus                          10         10         10         10   
Acer_Palmatum                        10         10         10         10   
...                                 ...        ...        ...        ...   
Tilia_Tomentosa                      10         10         10         10   
Ulmus_Bergmanniana                   10         10         10         10   
Viburnum_Tinus                       10         10         10         10   
Viburnum_x_Rhytidophylloides         10         10         10         10   
Zelkova_Serrata                      10        

On voit que pour chaque espèce de plantes, on a exactement 10 observations.

In [48]:
data.describe()

margin1     margin2     margin3     margin4     margin5     margin6  \
count  990.000000  990.000000  990.000000  990.000000  990.000000  990.000000   
mean     0.017412    0.028539    0.031988    0.023280    0.014264    0.038579   
std      0.019739    0.038855    0.025847    0.028411    0.018390    0.052030   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.001953    0.001953    0.013672    0.005859    0.001953    0.000000   
50%      0.009766    0.011719    0.025391    0.013672    0.007812    0.015625   
75%      0.025391    0.041016    0.044922    0.029297    0.017578    0.056153   
max      0.087891    0.205080    0.156250    0.169920    0.111330    0.310550   

          margin7     margin8     margin9    margin10  ...   texture55  \
count  990.000000  990.000000  990.000000  990.000000  ...  990.000000   
mean     0.019202    0.001083    0.007167    0.018639  ...    0.036501   
std      0.017511    0.002743    0.008933    0.016071  ...    0.063403   
min      0.000000    0.000000    0.000000    0.000000  ...    0.000000   
25%      0.005859    0.000000    0.001953    0.005859  ...    0.000000   
50%      0.015625    0.000000    0.005859    0.015625  ...    0.004883   
75%      0.029297    0.000000    0.007812    0.027344  ...    0.043701   
max      0.091797    0.031250    0.076172    0.097656  ...    0.429690   

        texture56   texture57   texture58   texture59   texture60   texture61  \
count  990.000000  990.000000  990.000000  990.000000  990.000000  990.000000   
mean     0.005024    0.015944    0.011586    0.016108    0.014017    0.002688   
std      0.019321    0.023214    0.025040    0.015335    0.060151    0.011415   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000977    0.000000    0.004883    0.000000    0.000000   
50%      0.000000    0.005859    0.000977    0.012695    0.000000    0.000000   
75%      0.000000    0.022217    0.009766    0.021484    0.000000    0.000000   
max      0.202150    0.172850    0.200200    0.106450    0.578130    0.151370   

        texture62   texture63   texture64  
count  990.000000  990.000000  990.000000  
mean     0.020291    0.008989    0.019420  
std      0.039040    0.013791    0.022768  
min      0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000977  
50%      0.003906    0.002930    0.011719  
75%      0.023438    0.012695    0.029297  
max      0.375980    0.086914    0.141600  

[8 rows x 192 columns]

In [49]:
data.isna().sum().unique()

array([0], dtype=int64)

## Preprocessing des données

On centre-réduit les données.

In [51]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X_train)

X = data.drop(columns=["species"]).values
y = data["species"].values
X_validation = data_test.copy().values

X = scaler.transform(X)
X_validation = scaler.transform(X_validation)

On sépare nos données en un jeu d'entraînement et un jeu de test.

In [53]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25)

In [55]:
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(742, 192) (248, 192) (742,) (248,)


## Modélisation

### Classificateur naïf

Pour avoir une baseline, on va entraîner un classificateur des k plus proches voisins.

In [58]:
from sklearn.neighbors import KNeighborsClassifier
kNN = KNeighborsClassifier(n_neighbors = 11)
kNN.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=11)

In [59]:
y_kNN = kNN.predict(X_test)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


In [61]:
y_kNN.shape

(248,)

In [87]:
from sklearn.metrics import confusion_matrix, accuracy_score
confusion_mat = pd.DataFrame(confusion_matrix(y_test, y_kNN))
accuracy_kNN = accuracy_score(y_test, y_kNN)
print(f"Score du classificateur 11-NN : {accuracy_kNN.round(4)*100}%")

Score du classificateur 11-NN : 69.35%


En conclusion, notre baseline est de 69.35% d'accuracy.

### Classificateur SVM 

Commençons par tester plusieurs méthodes de Support Vector Machine de base.

In [100]:
from sklearn.svm import LinearSVC
def svc(multiclass):
    svc_ = LinearSVC(verbose=0, multi_class=multiclass)
    svc_.fit(X_train, y_train)
    y_svc_ = svc_.predict(X_test)
    accuracy = accuracy_score(y_test, y_svc_)
    print(f"Score du SVC multiclasse {multiclass} : {accuracy.round(4)*100}%")
    
    return y_svc_

In [101]:
y_svc_ovr = svc("ovr")
y_svc_ovo = svc("crammer_singer")

Score du SVC multiclasse ovr : 74.6%
Score du SVC multiclasse crammer_singer : 83.47%


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


On voit que les SVC ovr et crammer_singer sont meilleures que le kNN. Voyons si on peut encore faire mieux.

In [115]:
help(LinearSVC)

Help on class LinearSVC in module sklearn.svm._classes:

class LinearSVC(sklearn.linear_model._base.LinearClassifierMixin, sklearn.linear_model._base.SparseCoefMixin, sklearn.base.BaseEstimator)
 |  LinearSVC(penalty='l2', loss='squared_hinge', *, dual=True, tol=0.0001, C=1.0, multi_class='ovr', fit_intercept=True, intercept_scaling=1, class_weight=None, verbose=0, random_state=None, max_iter=1000)
 |  
 |  Linear Support Vector Classification.
 |  
 |  Similar to SVC with parameter kernel='linear', but implemented in terms of
 |  liblinear rather than libsvm, so it has more flexibility in the choice of
 |  penalties and loss functions and should scale better to large numbers of
 |  samples.
 |  
 |  This class supports both dense and sparse input and the multiclass support
 |  is handled according to a one-vs-the-rest scheme.
 |  
 |  Read more in the :ref:`User Guide <svm_classification>`.
 |  
 |  Parameters
 |  ----------
 |  penalty : {'l1', 'l2'}, default='l2'
 |      Specifies t

### Optimisation des classificateurs 

Pour optimiser notre SVC, on va procéder par validations croisées.

In [109]:
from sklearn.model_selection import GridSearchCV

# On instancie l'estimateur
estimator = LinearSVC()

# On collecte toutes les valeurs possibles des paramètres
params = {"C" : np.logspace(-3,3,7),
          "penalty" : ["l1", "l2"],
          "loss" : ["hinge", "squared_hinge"],
          "dual" : [False],
          "multi_class" : ["ovr", "crammer_singer"]}

# On crée la grille de validation croisée
grid = GridSearchCV(estimator,
                   param_grid = params,
                   scoring = "accuracy",
                   n_jobs = -1,
                   cv = 8,
                   verbose = 1)
grid.fit(X_train, y_train)

Fitting 8 folds for each of 56 candidates, totalling 448 fits


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:676: UserWarning: The least populated class in y has only 4 members, which is less than n_splits=8.
  warnings.warn(
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
112 fits failed out of a total of 448.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
56 fits failed with the following error:
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\ProgramData\Anaconda3\lib\site-packages\sklearn\svm\_classes.py", line 257, in fit
    self.coef_, self

GridSearchCV(cv=8, estimator=LinearSVC(), n_jobs=-1,
             param_grid={'C': array([1.e-03, 1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02, 1.e+03]),
                         'dual': [False], 'loss': ['hinge', 'squared_hinge'],
                         'multi_class': ['ovr', 'crammer_singer'],
                         'penalty': ['l1', 'l2']},
             scoring='accuracy', verbose=1)

In [110]:
grid.best_params_

{'C': 100.0,
 'dual': False,
 'loss': 'hinge',
 'multi_class': 'crammer_singer',
 'penalty': 'l1'}

In [121]:
results = pd.DataFrame(grid.cv_results_)
results = results.drop(columns = [i for i in results.columns if "split" in i]).drop(columns = ["params"]).sort_values("rank_test_score")
results.head(20)

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
55      13.144056      1.856353         0.003907        0.006767  1000.0   
54      14.403905      1.873809         0.002376        0.001653  1000.0   
43      12.482315      1.432313         0.001251        0.000432   100.0   
46      13.043231      1.338060         0.001126        0.000332   100.0   
47      12.704581      1.242929         0.001876        0.000332   100.0   
42      12.660453      1.638149         0.001626        0.000484   100.0   
51      12.962851      1.495682         0.001501        0.000501  1000.0   
50      13.241497      1.394284         0.001501        0.000500  1000.0   
53       7.493438      0.435399         0.002126        0.000600  1000.0   
45       1.798759      0.462088         0.003876        0.005348   100.0   
52      15.061094      0.419762         0.002376        0.000697  1000.0   
37       1.589495      0.319112         0.003501        0.004823    10.0   
44       8.827525      0.194075         0.001876        0.000600   100.0   
39       6.416982      0.611579         0.001625        0.000484    10.0   
38       6.506489      0.633043         0.001503        0.000502    10.0   
35       6.549117      0.469728         0.001501        0.000501    10.0   
34       7.520598      0.521979         0.001126        0.000330    10.0   
36       5.545416      0.246820         0.001751        0.000662    10.0   
26       3.705731      0.315950         0.000000        0.000000     1.0   
31       4.708272      0.332227         0.001953        0.005167     1.0   

   param_dual     param_loss param_multi_class param_penalty  mean_test_score  \
55      False  squared_hinge    crammer_singer            l2         0.948793   
54      False  squared_hinge    crammer_singer            l1         0.948793   
43      False          hinge    crammer_singer            l2         0.948793   
46      False  squared_hinge    crammer_singer            l1         0.948793   
47      False  squared_hinge    crammer_singer            l2         0.948793   
42      False          hinge    crammer_singer            l1         0.948793   
51      False          hinge    crammer_singer            l2         0.948793   
50      False          hinge    crammer_singer            l1         0.948793   
53      False  squared_hinge               ovr            l2         0.947405   
45      False  squared_hinge               ovr            l2         0.944746   
52      False  squared_hinge               ovr            l1         0.941985   
37      False  squared_hinge               ovr            l2         0.928515   
44      False  squared_hinge               ovr            l1         0.911013   
39      False  squared_hinge    crammer_singer            l2         0.906937   
38      False  squared_hinge    crammer_singer            l1         0.906937   
35      False          hinge    crammer_singer            l2         0.906937   
34      False          hinge    crammer_singer            l1         0.906937   
36      False  squared_hinge               ovr            l1         0.902948   
26      False          hinge    crammer_singer            l1         0.885344   
31      False  squared_hinge    crammer_singer            l2         0.885344   

    std_test_score  rank_test_score  
55        0.014993                1  
54        0.014993                1  
43        0.014993                1  
46        0.014993                1  
47        0.014993                1  
42        0.014993                1  
51        0.014993                1  
50        0.014993                1  
53        0.016666                9  
45        0.015627               10  
52        0.021702               11  
37        0.032891               12  
44        0.029479               13  
39        0.027204               14  
38        0.027204               14  
35        0.027204               14  
34        0.027204               14  
36        0.025874   

Fort de ces résultats, on va entraîner deux nouveaux classificateurs :
- Un qui suit la méthode One-vs-Rest
- Un qui suit la méthode Crammer & Singer

et on va les tester sur le jeu de test.

In [130]:
svc_ovr = LinearSVC(dual=False, penalty="l2", loss="squared_hinge", C = 5000, multi_class="ovr")
svc_ovr.fit(X_train, y_train)
y_pred_ovr = svc_ovr.predict(X_test)
accuracy_ovr = accuracy_score(y_test, y_pred_ovr)
print(f"Score du SVC ovr : {accuracy_ovr.round(4)*100}%")

Score du SVC ovr : 92.34%


In [132]:
svc_crammer_singer = LinearSVC(C = 10000, multi_class="crammer_singer")
svc_crammer_singer.fit(X_train, y_train)
y_pred_crammer_singer = svc_crammer_singer.predict(X_test)
accuracy_crammer_singer = accuracy_score(y_test, y_pred_crammer_singer)
print(f"Score du SVC crammer singer : {accuracy_crammer_singer.round(4)*100}%")

Score du SVC crammer singer : 90.73%


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\svm\_base.py:1206: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


C'est sans appel, le SVC ovr est meilleur que le svc crammer singer. Cependant, en tâtonnant un peu, on se rend compte que l'on avait mal géré les valeurs de C. Refaisons une validation croisée plus précise.

In [133]:
# On instancie l'estimateur
estimator = LinearSVC(multi_class="ovr", dual=False)

# On collecte toutes les valeurs possibles des paramètres
params = {"C" : np.logspace(2, 4, 10),
          "penalty" : ["l1", "l2"],
          "loss" : ["hinge", "squared_hinge"]}

# On crée la grille de validation croisée
grid = GridSearchCV(estimator,
                   param_grid = params,
                   scoring = "accuracy",
                   n_jobs = -1,
                   cv = 8,
                   verbose = 1)
grid.fit(X_train, y_train)

Fitting 8 folds for each of 40 candidates, totalling 320 fits


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:676: UserWarning: The least populated class in y has only 4 members, which is less than n_splits=8.
  warnings.warn(
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
160 fits failed out of a total of 320.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
80 fits failed with the following error:
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\ProgramData\Anaconda3\lib\site-packages\sklearn\svm\_classes.py", line 257, in fit
    self.coef_, self

GridSearchCV(cv=8, estimator=LinearSVC(dual=False), n_jobs=-1,
             param_grid={'C': array([  100.        ,   166.81005372,   278.25594022,   464.15888336,
         774.26368268,  1291.54966501,  2154.43469003,  3593.8136638 ,
        5994.84250319, 10000.        ]),
                         'loss': ['hinge', 'squared_hinge'],
                         'penalty': ['l1', 'l2']},
             scoring='accuracy', verbose=1)

In [134]:
grid.best_params_

{'C': 2154.4346900318824, 'loss': 'squared_hinge', 'penalty': 'l1'}

In [138]:
results = pd.DataFrame(grid.cv_results_)
results.drop(columns= [i for i in results.columns if "split" in i or "params" in i]).sort_values("rank_test_score")

mean_fit_time  std_fit_time  mean_score_time  std_score_time      param_C  \
26      20.775519      0.680829         0.005283        0.009904   2154.43469   
34      24.576707      1.297229         0.003907        0.006768  5994.842503   
38      28.340716      1.044101         0.003905        0.006764      10000.0   
30      26.102233      1.134269         0.001953        0.005168  3593.813664   
7        5.960277      0.276260         0.002501        0.001119   166.810054   
19       7.367803      0.614506         0.002000        0.000708   774.263683   
23       8.781134      1.174209         0.005859        0.007564  1291.549665   
15       5.922822      1.178416         0.002251        0.000829   464.158883   
11       6.812674      0.940984         0.003981        0.003227    278.25594   
3        5.466245      0.132033         0.004629        0.005829        100.0   
35       9.413611      1.670670         0.000500        0.000708  5994.842503   
31      11.609897      1.311095         0.001953        0.005168  3593.813664   
27      13.482903      1.133769         0.001502        0.001002   2154.43469   
39       6.834863      1.217823         0.000000        0.000000      10000.0   
22      15.818064      0.284087         0.002253        0.000663  1291.549665   
18      14.525300      0.607030         0.002252        0.000661   774.263683   
14      16.543982      2.290588         0.002877        0.001166   464.158883   
6       11.696297      0.332313         0.003106        0.002335   166.810054   
10      14.624554      1.331485         0.004248        0.001559    278.25594   
2        8.814379      0.128738         0.019601        0.004530        100.0   
37       0.004085      0.006678         0.000000        0.000000      10000.0   
36       0.009766      0.007565         0.000000        0.000000      10000.0   
1        0.003906      0.006766         0.000000        0.000000        100.0   
33       0.000000      0.000000         0.000000        0.000000  5994.842503   
32       0.009766      0.007565         0.000000        0.000000  5994.842503   
4        0.014374      0.016695         0.000000        0.000000   166.810054   
5        0.006627      0.001218         0.000000        0.000000   166.810054   
29       0.013002      0.007450         0.000000        0.000000  3593.813664   
8        0.016873      0.007994         0.000000        0.000000    278.25594   
13       0.013752      0.012205         0.000000        0.000000   464.158883   
25       0.013672      0.005168         0.000000        0.000000   2154.43469   
24       0.007851      0.007851         0.000000        0.000000   2154.43469   
9        0.007751      0.003229         0.000000        0.000000    278.25594   
21       0.008372      0.001867         0.000000        0.000000  1291.549665   
20       0.008876      0.004076         0.000000        0.000000  1291.549665   
17       0.009625      0.006889         0.000000        0.000000   774.263683   
16       0.009499      0.005027         0.000000        0.000000   774.263683   
12       0.007377      0.002234         0.000000        0.000000   464.158883   
28       0.010543      0.006588         0.000000        0.000000  3593.813664   
0        0.005273      0.002781         0.000000        0.000000        100.0   

       param_loss param_penalty  mean_test_score  std_test_score  \
26  squared_hinge            l1         0.958173        0.017573   
34  squared_hinge            l1         0.956829        0.018769   
38  squared_hinge            l1         0.955484        0.012784   
30  squared_hinge            l1         0.951452        0.013356   
7   squared_hinge            l2         0.948779        0.015017   
19  squared_hinge            l2         0.947405        0.016666   
23  squared_hinge            l2         0.947405        0.016666   
15  squared_hinge            l2         0.947405        0.016666   
11  squared_hinge            l2         0.946076        0.016229   
3 

On voit qu'on a les meilleurs résultats lorsque la loss est la squared_hinge, C est dans [2200, 3600, 6000] et la penalty = l1. On va donc entraîner ces modèles.

In [145]:
def f(C_value):
    svc_ovr = LinearSVC(dual=False, penalty="l2", loss="squared_hinge", C = C_value, multi_class="ovr")
    svc_ovr.fit(X_train, y_train)
    y_pred_ovr = svc_ovr.predict(X_test)
    accuracy_ovr = accuracy_score(y_test, y_pred_ovr)
    print(f"Score du SVC ovr avec C = {C_value} : {accuracy_ovr.round(6)*100}%")

In [149]:
f(2000)

Score du SVC ovr avec C = 2000 : 91.9355%


In [147]:
f(3600)

Score du SVC ovr avec C = 3600 : 92.33869999999999%


In [148]:
f(6000)

Score du SVC ovr avec C = 6000 : 92.33869999999999%


Pour finir, on va donc prendre C = 2000, loss = "squared_hinge", penalty="l2", dual=False, multi_class="ovr".

## Conclusion

En guise de conclusion, on va prédire sur le jeu de validation les étiquettes.

In [152]:
X_validation

array([[ 1.95310000e-02,  9.76600000e-03,  7.81250000e-02, ...,
         1.18423789e-16,  3.90600000e-03,  5.37110000e-02],
       [ 7.81200000e-03,  5.85900000e-03,  6.44530000e-02, ...,
         9.77000000e-04,  3.71090000e-02,  4.49220000e-02],
       [ 4.12688963e-17, -9.68921912e-17,  1.95300000e-03, ...,
         1.56250000e-02, -1.79429984e-17,  4.12688963e-17],
       ...,
       [ 1.75780000e-02,  2.92970000e-02,  1.56250000e-02, ...,
         1.18423789e-16,  4.29690000e-02,  6.83600000e-03],
       [ 1.36720000e-02,  9.76600000e-03,  6.05470000e-02, ...,
         1.18423789e-16,  1.17190000e-02,  1.85550000e-02],
       [ 4.12688963e-17,  1.17190000e-01,  1.47132587e-16, ...,
         1.56250000e-02, -1.79429984e-17,  1.75780000e-02]])

In [153]:
model = LinearSVC(dual=False, penalty="l2", loss="squared_hinge", C = 2000, multi_class="ovr")
model.fit(X, y)
y_validation_pred = model.predict(X_validation)

In [154]:
accuracy_score(y, model.predict(X))

1.0

In [175]:
X_0 = data.loc[data.index==456].drop(columns=["species"]).values
y_0 = data.loc[data.index==456].species.values

print(y_0, model.predict(X_0))

['Prunus_Avium'] ['Prunus_Avium']


Bon, ça marche nickel ! Par contre, je me demande dans quelle mesure j'ai pas fait de surapprentrissage ?